### DISCUSSION TITLE
“dynamic” relationship in asyncio

### DISCUSSION THREAD

**Discussion by cancan101** [sqlalchemy/sqlalchemy#7039] [Usage Questions] [answered]
The [docs for asyncio](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession) say:
>The “dynamic” relationship loader strategy described at Dynamic Relationship Loaders is not compatible with the asyncio approach and cannot be used, unless invoked within the AsyncSession.run_sync() method described at Running Synchronous Methods and Functions under asyncio.

however that section does not have an explicit example of using the `AsyncSession.run_sync()` with an (existing) dynamic relationship.

**Comment by CaselIT** [comment]
Hi,
Following that link takes you the an example usage of run_sync: https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#session-run-sync
Inside of run_sync one is able to use the normal sqlalchemy interface, so the dynamic relationship can be used normally as illustrated at https://docs.sqlalchemy.org/en/14/orm/collections.html#dynamic-relationship

**Reply by cancan101** [reply]
Were there any thoughts around supporting dynamic queries in asyncio using the new select interface? For example something like:
```python
result = await session.execute(author.books)
```
where `author.books` is a dynamic relationship and is of the same type as returned by `select`?

It would be nice to be able to use the same syntax for defining and using dynamic relationships as for non dynamic relationships and not have to resort to using `with_parent` for one of them.

While the alternative of continuing to use `AsyncSession.run_sync()` is helpful as a stopgap, it would also be nice to be able to have a consistent interface to produce an executable select statement between dynamic relationships and and regular selects.

**Reply by zzzeek** [reply]
"author.books" is not clear in what it means so we prefer to keep the things passed to session.execute() crystal clear in what is going on.   2.0's emphasis is on clarity and fewer choices for argument styles.  many years of looking for "less typing" ways of doing things, for fear people would complain about verbosity, is what led us to have so many API issues in the first place, methods that accept half a dozen different argument styles which nobody understands, etc.

**Reply by zzzeek** [reply]
in particular i think this is quite clear that we are running a SELECT statement:

    result = await session.execute(select(Book).where(with_parent(author, Author.books)))

a key reason people like asyncio is that the decision has been made that "implicit IO" is bad.   so from my pov we should not be trying to re-introduce ways that IO or what kind of statements are being invoked would be less apparent.

if we wanted to slightly reduce verbosity, a helper on Author.books perhaps:

    result = await session.execute(Author.books.select_from(author))

shorter, but still the word "SELECT" is apparent.

**Reply by CaselIT** [reply]
Maybe for v2 we could deprecate `dynamic` and have something that replaces it what works the same in both sync and async.
Like it could be:
```py
# sync
result = session.scalars(author.book.select().where(Book.id > 42))
# async
result = await session.scalars(author.book.select().where(Book.id > 42))
```

The issue is what to name this mode, since 'select' is taken and 'query' would be a poor choice since it uses `select` not `query`

**Reply by zzzeek** [reply]
dynamic essentially already does that actually, we can just document that if you want:

    result = await session.execute(author.book.statement.where(Book.id > 42))

magic !

**Reply by CaselIT** [reply]
> magic !

Indeed.

maybe in v2 we could just wrap the object returned by dynamic in a proxy class that deprecates all attributes but `statement` and sometime later just have it return `.statement`

**Reply by CaselIT** [reply]
btw, mike you should probably start a new thread with that answer so we can select it as the answer

**Reply by cancan101** [reply]
I am all for not hiding implicit io ;-) and yes, any of these seem reasonable:
```python
result = await session.execute(author.books.statement)
result = await session.execute(author.books.select())
result = await session.execute(Author.books.select_from(author))
```
though one point of clarification, how would `books` be defined on the model?

**Reply by zzzeek** [reply]
at the moment it's the same old "lazy='dynamic'" :)   we can make a new lazy called "explicit_only" or something like that which is more modern internally

**Comment by CaselIT** [comment, marked as answer]
dynamic essentially already does that actually, we can just document that if you want:

```py
result = await session.execute(author.book.statement.where(Book.id > 42))
```

magic !

Copied from Mikes reply https://github.com/sqlalchemy/sqlalchemy/discussions/7039#discussioncomment-1347776
